tutorials/provisional/mosaic.py

```diff
@@ -0,0 +1,295 @@
+"""
+=======================================
+Complex and semantic figure composition
+=======================================
+
+.. warning::
+
+   This tutorial documents experimental / provisional API.
+   We are releasing this in v3.3 to get user feedback.  We may
+   make breaking changes in future versions with no warning.
+
+
+Laying out Axes in a Figure in a non uniform grid can be both tedious
+and verbose.  For dense even grids we have `.Figure.subplots` but for
+more complex layouts, such as Axes that span multiple columns / rows
+of the layout or leave some areas of the Figure blank you can use
+`.gridspec.GridSpec` (see :doc:`/tutorials/intermediate/gridspec`) or
+manually place your axes.  `.Figure.subplot_mosaic` aims to provide an
+interface to visually layout your axes (as either ASCII art or nested
+lists) to streamline this process.
+
+As a side effect of this interface, it is very natural to name your
+axes!  `.Figure.subplot_mosaic` returns a dictionary keyed on the
+labels used to layout the Figure.  By returning data structures with
+names it is easier to write plotting code that is independent of the
+Figure layout.
+
+
+This is inspired by a `proposed MEP
+<https://github.com/matplotlib/matplotlib/pull/4384>`__ and the
+`patchwork <https://github.com/thomasp85/patchwork>`__ library for R.
+While we do not implement the operator overloading style, we do
+provide a Pythonic API for specifying (nested) Axes layouts.
+
+"""
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+def label_axes(ax_dict):
+    """
+    Helper to set the title on the Axes.
+
+    Parameters
+    ----------
+    ax_dict : Dict[str, Axes]
+        Mapping between the title / label and the Axes.  Set the left
+        title to be equal to the key.
+    """
+    for k, ax in ax_dict.items():
+        ax.set_title(k, loc="left")
+
+
+###############################################################################
+# String specification
+# ====================
+#
+# Using single letters we can "draw" the Axes we want.  The following
+
+
+layout = """
+AB
+CD
+"""
```

Comment:
```suggestion
layout = """
    AB
    CD
    """
```
Should be indented as we'd do in code.